File: KBSVM_Classifiers.py
import numpy as np
import pandas as pd
import math
import random
import cvxpy as cp

def confusion_matrix1(y_true, y_pred):

    x = list(filter(None.__ne__, set(y_true)))

    a = [[0 for x in range(2)] for y in range(2)]
    for j in range(len(x)):
        s1, s2, k = 0, 0, 0
        for i in range(len(y_true)):
            if ((x[j] == y_true[i]) & (y_true[i] == y_pred[i])):
                s1 += 1
            elif ((x[j] == y_true[i]) & (y_true[i] != y_pred[i])):
                s2 += 1

        a[j][j] = s1
        a[j][~j] = s2
    a = np.array(a).reshape((2, 2))
    print("CONFUSION MATRIX :: ")
    print(a)
    sum = 0
    for i in range(len(a)):
        sum += a[i, i]
    error = 1 - (sum / len(y_true))
    return error

def cross_valid(training_data_example, training_data_label):

    #---------------------------------------------K-FOLD CROSS-VALIDATION----------------------------------------------
    # Length of the training dataset : 500
    n1 = len(training_data_label)
    attr = len(training_data_example[0, :])
    # Number of folds can be changed by changing the kfold value....................................................
    kfold = 10;
    div = math.floor(n1 / kfold);
    # Assigning different hyper-parameter values
    Cdata_val=[]
    Cadv_val=[]
    # Dictionary to store (Cdata, Cadv) pair........................................................................
    dic={}
    count=0
    # Number of different Cdata & Cadv value........................................................................
    n_c=8
    for i in range(n_c):
        Cdata_val.append(math.pow(10,(-4+i)))
        Cadv_val.append(math.pow(10,(-4+i)))

    #--------------------------------------------VARIABLES -------------------------------------------------------------
    w = cp.Variable(shape=(attr, 1))
    eta = cp.Variable((attr, 1))
    zeta = cp.Variable((1, 1))
    b = cp.Variable((1, 1))
    u1 = cp.Variable((2, 1))
    u2 = cp.Variable((2, 1))
    u3 = cp.Variable((1,1))
    # -----------------------------------------ADVICE SET----------------------------------------------------------------
    # Advice set 1...........
    l = [0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0]
    D1 = np.array(l)
    D1 = np.reshape(D1, (2, attr))
    l = [100, 25]
    d1 = np.array(l)
    d1 = np.reshape(d1, (2, 1))
    # Advice set 2...........
    l = [0, -1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, -1, 0, 0]
    D2 = np.array(l)
    D2 = np.reshape(D2, (2, attr))
    l = [-126, -30]
    d2 = np.array(l)
    d2 = np.reshape(d2, (2, 1))
    # Advice set 3............
    l = [0, 0, 0, 0, 0, 0, 0, -1]
    D3 = np.array(l)
    D3 = np.reshape(D3, (1, attr))
    l = [-45]
    d3 = np.array(l)
    d3 = np.reshape(d3, (1, 1))
    # -----------------------------INDEXES FOR CROSS-VALIDATION------------------------------------------------------------
    cv_test_idx=[] #---CV TEST DATA SET
    cv_train_idx=[] #---CV TRAIN DATA SET
    for i in range(kfold):
        hold = random.sample(range(n1), div)
        cv_test_idx.append(list(hold))
        cv_train_idx.append(list(np.setdiff1d(range(n1),hold)))
    #---- PARAMETERS FOR CROSS-VALIDATION--------------------------------------------------------------------------------
    n2 = len(cv_train_idx[0]);
    xi = cp.Variable((n2, 1));
    Error = np.zeros((kfold, n_c, n_c))
    # ---------------------------------- OBJECTIVE & CONSTRAINTS TERMS --------------------------------------------------
    term12 = cp.abs(w)
    term1 = cp.sum(term12)
    term31 = cp.sum(cp.abs(eta))
    term32 = cp.sum(zeta)
    c2 = -w + (D1.T * u1) + eta
    c3 = b - 1 - (d1.T * u1) + zeta
    c4 = w + (D2.T * u2) + eta
    c5 = -b - 1 - (d2.T * u2) + zeta
    c6 = w + (D3.T * u3) + eta
    c7 = -b - 1 - (d3.T * u3) + zeta

    #--------------------------------- CROSS-VALIDATION -----------------------------------------------------------------
    for i in range(len(Cdata_val)):
        Cdata = Cdata_val[i]
        for j in range(len(Cadv_val)):
            Cadv = Cadv_val[j]
            #
            # Dictionary....
            key= count
            dic.setdefault(key, ())
            dic[key]=(Cdata,Cadv)
            #
            #
            for k in range(kfold):
                #
                #
                x = training_data_example[cv_train_idx[k],:]
                y = training_data_label[cv_train_idx[k]]
                c1 = np.diag(y) * (x * w + b) - 1 + xi
                term2 = Cdata * cp.sum(xi)
                term3 = Cadv * (term31 + term32)
                Objective = cp.Minimize(term1 + term2 + term3)
                Constraints = [xi >= 0, u1 >= 0, u2 >= 0, zeta >= 0, c1 >= 0, c2 == 0, c3 >= 0, c4 == 0, c5 >= 0]
                Prob = cp.Problem(Objective, Constraints)
                x_cv_test = training_data_example[cv_test_idx[k],:]
                y_cv_test = training_data_label[cv_test_idx[k]]
                Prob.solve(solver=cp.GLPK, verbose=True)
                # OPTIMIZED VALUES OF w & b ----------------
                #
                w_Opt = np.array(w.value, dtype=float)
                b_Opt = np.array(b.value, dtype=float)
                #
                #
                y_cv_pred = np.sign(np.matmul(x_cv_test, w_Opt) + b_Opt);
                y_cv_pred = np.reshape(y_cv_pred,(1,len(y_cv_pred)))
                y_cv_pred = y_cv_pred[0]
                #
                #
                Err = confusion_matrix1(y_cv_test, y_cv_pred)
                Error[k,i,j] = Err
            #
            count+=1
            #

    # ERROR OF ALL (Cdata, Cadv) PAIR AFTER KFOLD CV IN Error.txt FILE ---------------------------------
    #
    with open('Error.txt', 'w') as outfile:
        for j in Error:
            j=np.reshape(j,(1,n_c * n_c))
            np.savetxt(outfile, j)
    #
    # MEAN ERROR OF ALL (Cdata, Cadv) PAIR AFTER KFOLD CV IN Mean_Error.txt FILE ------------------------
    err = np.mean(Error, axis=0)
    err=np.reshape(err,(1,n_c * n_c))
    with open('Mean_Error.txt', 'w') as outfile:
        np.savetxt(outfile, err)
    #
    # RETURNING THE BEST (Cdata, Cadv) PAIR --------------------------------------------------------------
    err1=err[0]
    min_error= np.min(err1)
    idx1= np.where(err1 == min_error)
    idx=idx1[0]
    hold = dic[idx[0]]
    min_Cdata = hold[0]
    min_Cadv = hold[1]
    #
    #
    return (min_Cdata, min_Cadv)

def train_svm_with_data_and_advice(training_data_example, training_data_label, test_data_example, test_data_label):

    #
    # --------------------------------------CROSS-VALIDATION------------------------------------------------
    # BEST (Cdata, Cadv) PAIR ...................
    Cdata, Cadv = cross_valid(training_data_example, training_data_label)

    #Cdata, Cadv = 0.1, 0.1
    #
    #
    # --------------------------------------------VARIABLES -------------------------------------------------------------
    n1 = len(training_data_label)
    n2 = len(test_data_label)
    attr = len(training_data_example[0,:])
    idx_glucose = 1
    idx_BMI = 5
    w = cp.Variable(shape=(attr,1))
    xi = cp.Variable((n1,1))
    eta = cp.Variable((attr,1))
    zeta = cp.Variable((1,1))
    x = training_data_example
    y = training_data_label
    b = cp.Variable((1,1))
    u1 = cp.Variable((2,1))
    u2 = cp.Variable((2,1))
    u3 = cp.Variable((1,1))

    #-----------------------------------------ADVICE SET----------------------------------------------------------------
    # Advice set 1 (Negative Advice)...........
    l=[0,1,0,0,0,0,0,0, 0,0,0,0,0,1,0,0]
    D1 = np.array(l)
    D1=np.reshape(D1,(2,attr))
    l = [100, 25]
    d1 = np.array(l)
    d1=np.reshape(d1,(2,1))
    # Advice set 2 (Positive Advice)...........
    l = [0, -1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, -1, 0, 0]
    D2 = np.array(l)
    D2 = np.reshape(D2, (2, attr))
    l = [-126, -30]
    d2 = np.array(l)
    d2 = np.reshape(d2, (2, 1))
    # Advice set 3 (Positive Advice)............
    l = [0, 0, 0, 0, 0, 0, 0, -1]
    D3 = np.array(l)
    D3 = np.reshape(D3, (1, attr))
    l = [-45]
    d3 = np.array(l)
    d3 = np.reshape(d3, (1, 1))
    #----------------------------------- OBJECTIVE & CONSTRAINT TERMS -------------------------------------------------
    term12=cp.abs(w)
    term1 = cp.sum(term12)
    #term1 = 0.5 * (w.T * w)
    term2 = Cdata * cp.sum(xi)
    term31 = cp.sum(cp.abs(eta))
    term32 = cp.sum(zeta)
    term3 = Cadv * (term31 + term32)
    Objective = cp.Minimize(term1 + term2 + term3)
    c1 = np.diag(y) * (x*w+b) - 1 + xi
    c2 = -w + (D1.T * u1) + eta
    c3 = b -1 - (d1.T * u1) + zeta
    c4 = w + (D2.T * u2) + eta
    c5 = -b - 1 - (d2.T * u2) + zeta
    c6 = w + (D3.T * u3) + eta
    c7 = -b - 1 - (d3.T * u3) + zeta
    Constraints = [xi >= 0, u1 >= 0, u2>=0, zeta >= 0 , c1 >= 0, c2 == 0, c3 >= 0, c4 == 0, c5 >= 0]
    Prob = cp.Problem(Objective, Constraints)
    Prob.solve(solver=cp.GLPK, verbose=True)
    #
    #------ OPTIMIZED VALUES OF w & b ----------------
    w_Opt = np.array(w.value, dtype = float)
    b_Opt = np.array(b.value, dtype = float)
    print("\n-----------------------------------------------\n ------::: PROBLEM STATUS :::------\n",Prob.status)
    print("\n OPTIMIZED VALUE = ",Prob.value,'\n')
    print("The best Cdata, Cadv pair is = ", "( ",Cdata, " , ",Cadv," )" "\n")
    #---- PREDICTED LABELS OF THE TEST DATA ---------------------------------------
    x_test = test_data_example
    y_test = test_data_label
    y_pred = np.sign(np.matmul(x_test, w_Opt) + b_Opt)
    y_pred = np.reshape(y_pred, (1, len(y_pred)))
    y_pred = y_pred[0]
    #
    #
    Error = confusion_matrix1(y_test, y_pred)
    print("\n Error on the test data set = ",Error)
    

if __name__ == '__main__':

    #--------------------------------------------------------------------------------------------------------------------------------------

    training_data = pd.read_csv('diabetes_dataset/training_data.csv',header=None)
    training_data=np.array(training_data)
    training_data_example = training_data[:, 0:8]
    training_data_label = training_data[:, 8]
    test_data = pd.read_csv('diabetes_dataset/test_data.csv',header=None)
    test_data = np.array(test_data)
    test_data_example = test_data[:, 0:8]
    test_data_label = test_data[:, 8]
    train_svm_with_data_and_advice(training_data_example, training_data_label, test_data_example, test_data_label)

File: KBSVM_Classifiers.m
clear;
close all;
%{
%
% READ THE DIABETES.CSV FILE
dataset= readmatrix('diabetes.csv');
data_samples=length(dataset(:,1));
dataset(dataset(:,9)==0,9)=-1;
% 500 DATA SAMPLES AS TRAINING SAMPLES
training_samples=500;
% 268 DATA SAMPLES AS TEST SAMPLES
test_samples=268;
% SAMPLING THE INDEX FOR GENERATING TRAINING & TEST DATA
population=1:data_samples;
training_index=randsample(population,training_samples);
test_index=setdiff(population, training_index);
% Replace this s_train_data with training_data :::::::::
%s_train_data=dataset(training_index,:);
training_data=dataset(training_index,:);
writematrix(training_data,'training_data.csv');
test_data=dataset(test_index,:);
writematrix(test_data,'test_data.csv');
training_data=[];
test_data=[];
%}
training_data= readmatrix('training_data.csv');
test_data= readmatrix('test_data.csv');
%
% 10-FOLD CROSS-VALIDATION ::::::::::::::
%
% Length of the training dataset : 500
n1=length(training_data(:,9));
label=unique(training_data(:,9));
% Number of folds can be changed by changing the kfold value
kfold=1;
div = floor(n1/kfold);
% Assigning different hyper-parameter values
%Cdata= 10.^(-3:3);
%Cadv= 10.^(-3:3);
% Hyper-parameters after Cross-Validation
Cdata= 0.1;
Cadv= 0.1;
%
% Train a SVM classifier on the data & advice ....
%
% Advice set 1 for negetive examples....
D1 = [0 1 0 0 0 0 0 0; 0 0 0 0 0 1 0 0];
d1 = [100; 25];
% Advice set 2 for positive examples....
D2 = [0 -1 0 0 0 0 0 0; 0 0 0 0 0 -1 0 0];
d2 = [-126; -30];
%
%
u1 = sdpvar(2, 1);
u2 = sdpvar(2, 1);
eta = sdpvar(8, 1);
zeta = sdpvar(1, 1);
w = sdpvar(8, 1);
b = sdpvar(1, 1);
%
%
% Randomly dividing the Training Dataset into training data & test data for CV
population=1:n1;
for i = 1:kfold
    cv_test_idx{i}=randsample(population,div);
    cv_train_idx{i}=setdiff(population, div);
    
end
%
% Length of the CV training dataset : n2
n2=length(training_data(cv_train_idx{1}, 1:8 ));
xi = sdpvar(n2, 1);
% K-Fold Cross-Validation
for i = Cdata
    for j = Cadv
        for k = 1:kfold
            % Training data for CV
            x= training_data(cv_train_idx{k},1:8);
            y= training_data(cv_train_idx{k}, 9);
            % Testing data for CV
            x_cv_test= training_data(cv_test_idx{k},1:8);
            y_cv_test= training_data(cv_test_idx{k}, 9);
            %
            Constraints_adv = [ diag(y) * (x*w + b) - 1 + xi >= 0, xi >= 0,...
                    u1 >= 0,u2 >= 0, zeta >= 0, -w + D1'*u1 + eta == 0,...
                    +b - 1 - d1'*u1 + zeta >= 0, w + D2'*u2 + eta == 0,...
                    -b - 1 - d2'*u2 + zeta >= 0];
            %
            Objective_adv = sum(abs(w)) + i * sum(xi) + j * (sum(abs(eta)) + sum(zeta));
            %
            diagnostic = optimize(Constraints_adv, Objective_adv, sdpsettings('solver', 'sedumi', 'verbose', 0));
            wOpt_adv = double(w);
            bOpt_adv = double(b);
            % Predicting labels of the CV test data
            y_cv_pred = sign(x_cv_test * wOpt_adv + bOpt_adv);
            % Confusion matrix for error calculation
            mtx=[];
            for l = 1: length(label(:,1))
            s1=0;
            s2=0;
                for l1 = 1:length(y_cv_test(:,1))
                    if ((label(l) == y_cv_test(l1)) && (y_cv_test(l1) == y_cv_pred(l1)))
                        s1 = s1+ 1;
                    elseif ((label(l) == y_cv_test(l1)) && (y_cv_test(l1) ~= y_cv_pred(l1)))
                        s2 = s2 + 1;
                    end
                end
                mtx= [mtx,s1];
                mtx=[mtx,s2];
            end
            %disp(mtx);
            C1= find(Cdata == i);
            C2= find(Cadv == j);
            % Matrix containing error for all possible combination...
            % of hyperparameter values for each fold 
            Error(k, C1, C2) = ((mtx(2)+mtx(4))/(length(y_cv_test(:,1))));

        end
    end
end

%writematrix(Error,'5-fold_CV.csv');
% Testing the classifier on the test data set
x=training_data(:,1:8);
y=training_data(:,9);
xi = sdpvar(n1, 1);
%
Constraints_adv = [ diag(y) * (x*w + b) - 1 + xi >= 0, xi >= 0,...
                    u1 >= 0, u2 >= 0, zeta >= 0, -w + D1'*u1 + eta == 0,...
                    +b - 1 - d1'*u1 + zeta >= 0, w + D2'*u2 + eta == 0,...
                    -b - 1 - d2'*u2 + zeta >= 0];

%Objective_adv = 0.5*(w'*w) + Cdata * sum(xi) + Cadv * (sum(abs(eta)) + sum(zeta));
Objective_adv = sum(abs(w)) + Cdata * sum(xi) + Cadv * (sum(abs(eta)) + sum(zeta));
%
diagnostic = optimize(Constraints_adv, Objective_adv, sdpsettings('solver', 'sedumi', 'verbose', 0));
wOpt_adv = double(w);
bOpt_adv = double(b);
disp(diagnostic);
disp(wOpt_adv);
disp(bOpt_adv);
fprintf('Objective value = %g.\n', double(Objective_adv));
%PREDICTED VALUE WITH ADVICE
x_test=test_data(:,1:8);
y_true=test_data(:,9);
y_pred=sign(x_test * wOpt_adv + bOpt_adv);
%disp(y_pred);
% Confusion matrix for error calculation
mtx=[];
for j = 1: length(label(:,1))
    s1=0;
    s2=0;
    for i = 1:length(y_true(:,1))
        if ((label(j) == y_true(i)) && (y_true(i) == y_pred(i)))
            s1 = s1+ 1;
        elseif ((label(j) == y_true(i)) && (y_true(i) ~= y_pred(i)))
            s2 = s2 + 1;
        end
   end
   mtx= [mtx,s1];
   mtx=[mtx,s2];
end
%disp(mtx);
Error=(mtx(2)+mtx(4))/(length(y_true(:,1)));
fprintf('\n Error on the test data set = %0.4f \n', Error)
